app/scripts/lib/rpc-method-middleware/handlers/send-metadata.ts

```diff
@@ -0,0 +1,91 @@
+import { ethErrors } from 'eth-rpc-errors';
+import {
+  JsonRpcEngineEndCallback,
+  JsonRpcEngineNextCallback,
+} from 'json-rpc-engine';
+import {
+  Json,
+  JsonRpcParams,
+  JsonRpcRequest,
+  PendingJsonRpcResponse,
+} from '@metamask/utils';
+import {
+  PermissionSubjectMetadata,
+  SubjectType,
+} from '@metamask/permission-controller';
+import { MESSAGE_TYPE } from '../../../../../shared/constants/app';
+import { HandlerWrapper } from './types';
+
+type SubjectMetadataToAdd = PermissionSubjectMetadata & {
+  name?: string | null;
+  subjectType?: SubjectType | null;
+  extensionId?: string | null;
+  iconUrl?: string | null;
+} & Record<string, Json>;
+
+type AddSubjectMetadata = (metadata: SubjectMetadataToAdd) => void;
+
+type SendMetadataOptions = {
+  addSubjectMetadata: AddSubjectMetadata;
+  subjectType: SubjectType;
+};
+
+type SendMetadataConstraint<Params extends JsonRpcParams = JsonRpcParams> = {
+  implementation: (
+    req: JsonRpcRequest<Params>,
+    res: PendingJsonRpcResponse<true>,
+    _next: JsonRpcEngineNextCallback,
+    end: JsonRpcEngineEndCallback,
+    { addSubjectMetadata, subjectType }: SendMetadataOptions,
+  ) => void;
+} & HandlerWrapper;
+/**
+ * This internal method is used by our external provider to send metadata about
+ * permission subjects so that we can e.g. display a proper name and icon in
+ * our UI.
+ */
+const sendMetadata = {
+  methodNames: [MESSAGE_TYPE.SEND_METADATA],
+  implementation: sendMetadataHandler,
+  hookNames: {
+    addSubjectMetadata: true,
+    subjectType: true,
+  },
+} satisfies SendMetadataConstraint;
+export default sendMetadata;
+
+/**
+ * @param req - The JSON-RPC request object.
+ * @param res - The JSON-RPC response object.
+ * @param _next - The json-rpc-engine 'next' callback.
+ * @param end - The json-rpc-engine 'end' callback.
+ * @param options
+ * @param options.addSubjectMetadata - A function that records subject
+ * metadata, bound to the requesting origin.
+ * @param options.subjectType - The type of the requesting origin / subject.
+ */
+function sendMetadataHandler<Params extends JsonRpcParams = JsonRpcParams>(
+  req: JsonRpcRequest<Params>,
+  res: PendingJsonRpcResponse<true>,
+  _next: JsonRpcEngineNextCallback,
+  end: JsonRpcEngineEndCallback,
+  { addSubjectMetadata, subjectType }: SendMetadataOptions,
+): void {
+  const { origin, params } = req;
+  if (params && typeof params === 'object' && !Array.isArray(params)) {
```

Comment:
`isObject` from `@metamask/utils` should be an equivalent check, but applying this gives us new type errors.

```suggestion
  if (isObject(params) {
```

The fact that `tsc` in CI is not raising these type errors tells us that the implicit `any` typing for `params` in this function is not just a local issue.